File: algorithms/python/230_kth_smallest_element_in_a_bst/kth_smallest_element_in_a_bst.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def kth_smallest(root: Optional[TreeNode], k: int) -> int:
    """
    中序遍历
    因为二叉搜索树和中序遍历的特性，二叉搜索树的中序遍历是按照键增加的顺序进行的。
    """
    stack = []
    node = root
    while node or stack:
        while node:
            stack.append(node)
            node = node.left
        node = stack.pop()
        k -= 1
        if k == 0:
            return node.val
        node = node.right

File: algorithms/230-kth-smallest-element-in-a-bst/kth-smallest-element-in-a-bst.go
package problem230

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// kthSmallest 中序遍历
func kthSmallest(root *TreeNode, k int) int {
	var stack []*TreeNode
	node := root
	for {
		for node != nil {
			stack = append(stack, node)
			node = node.Left
		}
		stack, node = stack[:len(stack)-1], stack[len(stack)-1]
		k--
		if k == 0 {
			return node.Val
		}
		node = node.Right
	}
}
